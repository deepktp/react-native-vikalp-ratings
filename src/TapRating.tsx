import React, { useState, useEffect } from 'react';

import {
  StyleSheet,
  Text,
  View,
  type ImageStyle,
  type StyleProp,
  type ViewStyle,
} from 'react-native';

import Star from './components/Star';

export type TapRatingProps = {
  /**
   * Total number of ratings to display.
   *
   * @default 5
   */
  count?: number;

  /**
   * Labels to show when each value is tapped.
   *
   * e.g. If the first star is tapped, then value in index 0 will be used as the label.
   *
   * @default ['Terrible', 'Bad', 'Okay', 'Good', 'Great']
   */
  reviews?: string[];

  /**
   * Determines if to show the reviews above the rating.
   *
   * @default true
   */
  showRating?: boolean;

  /**
   * Color value for review.
   *
   * @default #f1c40f
   */
  reviewColor?: string;

  /**
   * Size value for review.
   *
   * @default 40
   */
  reviewSize?: number;

  /**
   * Initial value for the rating
   *
   * @default 3
   */
  defaultRating?: number;

  /**
   * Style for star container
   *
   * @default undefined
   */
  starContainerStyle?: StyleProp<ViewStyle>;

  /**
   * Style for rating container
   *
   * @default undefined
   */
  ratingContainerStyle?: StyleProp<ViewStyle>;

  /**
   * Callback method when the user finishes rating. Gives you the final rating value as a whole number
   */
  onFinishRating?: (value: number) => void;

  /**
   * Whether the rating can be modified by the user
   *
   * @default false
   */
  isDisabled?: boolean;

  /**
   * Color value for filled stars.
   *
   * @default #004666
   */
  selectedColor?: string;

  /**
   * Color value for not filled stars.
   *
   * @default #BDC3C7
   */
  unSelectedColor?: string;

  /**
   * Size of rating image
   *
   * @default 40
   */
  size?: number;

  /**
   * Pass in a custom base image source
   */
  starImage?: string;

  /**
   * Style for star
   */
  starStyle?: ImageStyle;
};

const TapRating: React.FunctionComponent<TapRatingProps> = ({
  count = 5,
  reviews = ['Terrible', 'Bad', 'Okay', 'Good', 'Great'],
  showRating = true,
  reviewColor = 'rgba(230, 196, 46, 1)',
  reviewSize = 25,
  defaultRating = 3,
  starContainerStyle,
  ratingContainerStyle,
  onFinishRating,
  selectedColor,
  unSelectedColor,
  isDisabled = false,
  starImage,
  starStyle,
}) => {
  const [position, setPosition] = useState<number>(defaultRating);

  useEffect(() => {
    if (defaultRating === null || defaultRating === undefined) {
      setPosition(3);
    } else {
      setPosition(defaultRating);
    }
  }, [defaultRating]);

  const renderStars = (rating_array: React.ReactElement[]) => {
    return rating_array.map((star) => star);
  };

  const starSelectedInPosition = (position: number) => {
    if (typeof onFinishRating === 'function') {
      onFinishRating(position);
    }
    setPosition(position);
  };

  const rating_array = [];

  for (let index = 0; index < count; index++) {
    rating_array.push(
      <Star
        key={index}
        position={index + 1}
        starSelectedInPosition={starSelectedInPosition}
        fill={position >= index + 1}
        isDisabled={isDisabled}
        selectedColor={selectedColor}
        unSelectedColor={unSelectedColor}
        size={reviewSize}
        starImage={starImage}
        starStyle={starStyle}
      />
    );
  }

  return (
    <View
      style={StyleSheet.flatten([styles.ratingContainer, ratingContainerStyle])}
    >
      {showRating && (
        <Text
          style={[
            styles.reviewText,
            { fontSize: reviewSize, color: reviewColor },
          ]}
        >
          {reviews[position - 1]}
        </Text>
      )}
      <View
        style={StyleSheet.flatten([styles.starContainer, starContainerStyle])}
      >
        {renderStars(rating_array)}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  ratingContainer: {
    backgroundColor: 'transparent',
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
  },
  reviewText: {
    fontWeight: 'bold',
    margin: 10,
  },
  starContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

export default TapRating;
