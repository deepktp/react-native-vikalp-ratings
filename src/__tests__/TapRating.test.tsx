import { fireEvent, render } from '@testing-library/react-native';
import TapRating from '../TapRating';

describe('TapRating Aka AirbnbRating', () => {
  it('should export SwipeRating', () => {
    const TapRatingLocal = require('../TapRating');
    expect(TapRatingLocal).toBeDefined();
  });

  it('should match snapshot', () => {
    const { toJSON } = render(<TapRating />); //to ensure default props are not changed
    expect(toJSON()).toMatchSnapshot();
  });

  it('show rating true should show rating', () => {
    const { getByTestId } = render(<TapRating showRating={true} />);
    const ratingText = getByTestId('RNVUI__TapRating-showRating');
    expect(ratingText).toBeOnTheScreen();
  });

  it('number of stars should match count prop', () => {
    const { getAllByTestId } = render(<TapRating count={4} />);
    const ratingStar = getAllByTestId('RNVUI__Star');
    expect(ratingStar).toHaveLength(4);
  });

  it('should show reviews', () => {
    const defaultRating = 1;
    const ratingReviews = ['Terrible', 'Bad', 'Okay', 'Good', 'Great'];

    const { getByTestId } = render(
      <TapRating reviews={ratingReviews} defaultRating={defaultRating} />
    );
    const ratingText = getByTestId('RNVUI__TapRating-showRating');
    expect(ratingText).toBeOnTheScreen();
    expect(ratingText).toHaveTextContent(
      ratingReviews[defaultRating - 1] || ''
    );
  });

  it('review size should match reviewSize prop', () => {
    const reviewSize = 32;
    const { getByTestId } = render(<TapRating reviewSize={reviewSize} />);
    const ratingText = getByTestId('RNVUI__TapRating-showRating');
    expect(ratingText.props.style.fontSize).toBe(reviewSize);
  });

  it('rating image size should should match image size', () => {
    const imageSize = 32;
    const { getAllByTestId } = render(<TapRating size={imageSize} />);
    const ratingStars = getAllByTestId('RNVUI__Star-image');
    ratingStars.forEach((star) => {
      expect(star.props.style.width).toBe(imageSize);
      expect(star.props.style.height).toBe(imageSize);
    });
  });

  it('should call onFinishRating with the correct rating when rating is selected', () => {
    const pressableRating = 2;
    const onFinishRating = jest.fn();
    const { getAllByTestId } = render(
      <TapRating onFinishRating={onFinishRating} />
    );
    const ratingStars = getAllByTestId('RNVUI__Star');
    fireEvent.press(ratingStars[pressableRating - 1]);
    expect(onFinishRating).toHaveBeenCalledWith(pressableRating);
  });

  it('should not call onFinishRating when isDisabled is true', () => {
    const onFinishRating = jest.fn();
    const { getAllByTestId } = render(
      <TapRating isDisabled={true} onFinishRating={onFinishRating} />
    );
    const ratingStars = getAllByTestId('RNVUI__Star');
    fireEvent.press(ratingStars[0]);
    expect(onFinishRating).not.toHaveBeenCalled();
  });

  //image test
  it('should use custom star image', () => {
    const customStarImage = require('../images/bell.png');
    const { getAllByTestId } = render(
      <TapRating starImage={customStarImage} />
    );
    const ratingStars = getAllByTestId('RNVUI__Star-image');
    ratingStars.forEach((star) => {
      expect(star.props.source).toEqual(customStarImage);
    });
  });

  it('selected color should match selectedColor prop', () => {
    const selectedColor = 'red';
    const { getAllByTestId } = render(
      <TapRating
        selectedColor={selectedColor}
        count={5}
        defaultRating={5} //so that all stars are selected
      />
    );
    const ratingStars = getAllByTestId('RNVUI__Star-image');
    ratingStars.forEach((star) => {
      expect(star.props.style.tintColor).toBe(selectedColor);
    });
  });

  it('unselected color should match unSelectedColor prop', () => {
    const unSelectedColor = 'blue';
    const { getAllByTestId } = render(
      <TapRating
        unSelectedColor={unSelectedColor}
        count={5}
        defaultRating={0} //so that all stars are unselected
      />
    );
    const ratingStars = getAllByTestId('RNVUI__Star-image');
    ratingStars.forEach((star) => {
      expect(star.props.style.tintColor).toBe(unSelectedColor);
    });
  });
  it('should apply custom star style', () => {
    const customStarStyle = { backgroundColor: '#aabbcc' };
    const { getAllByTestId } = render(
      <TapRating starStyle={customStarStyle} />
    );
    const ratingStars = getAllByTestId('RNVUI__Star-image');
    ratingStars.forEach((star) => {
      expect(star.props.style.backgroundColor).toEqual(
        customStarStyle.backgroundColor
      );
    });
  });

  it('should apply custom rating container style', () => {
    const customContainerStyle = { backgroundColor: '#aabbcc' };
    const { getByTestId } = render(
      <TapRating ratingContainerStyle={customContainerStyle} />
    );
    const ratingContainer = getByTestId('RNVUI__TapRating');
    expect(ratingContainer.props.style.backgroundColor).toEqual(
      customContainerStyle.backgroundColor
    );
  });

  it('should apply custom star container style', () => {
    const customStarContainerStyle = { backgroundColor: '#aabbcc' };
    const { getByTestId } = render(
      <TapRating starContainerStyle={customStarContainerStyle} />
    );
    const ratingContainer = getByTestId('RNVUI__TapRating-starContainer');
    expect(ratingContainer.props.style.backgroundColor).toEqual(
      customStarContainerStyle.backgroundColor
    );
  });
});
