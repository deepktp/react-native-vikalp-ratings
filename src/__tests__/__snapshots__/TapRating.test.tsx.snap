// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TapRating Aka AirbnbRating should match snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
  testID="RNVUI__TapRating"
>
  <Text
    style={
      {
        "color": "rgba(230, 196, 46, 1)",
        "fontSize": 25,
        "fontWeight": "bold",
        "margin": 10,
      }
    }
    testID="RNVUI__TapRating-showRating"
  >
    Okay
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
    testID="RNVUI__TapRating-starContainer"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="RNVUI__Star"
    >
      <Image
        collapsable={false}
        source={
          {
            "testUri": "../../../src/images/airbnb-star.png",
          }
        }
        style={
          {
            "height": 40,
            "margin": 3,
            "tintColor": "#f1c40f",
            "transform": [
              {
                "scale": 1,
              },
            ],
            "width": 40,
          }
        }
        testID="RNVUI__Star-image"
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="RNVUI__Star"
    >
      <Image
        collapsable={false}
        source={
          {
            "testUri": "../../../src/images/airbnb-star.png",
          }
        }
        style={
          {
            "height": 40,
            "margin": 3,
            "tintColor": "#f1c40f",
            "transform": [
              {
                "scale": 1,
              },
            ],
            "width": 40,
          }
        }
        testID="RNVUI__Star-image"
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="RNVUI__Star"
    >
      <Image
        collapsable={false}
        source={
          {
            "testUri": "../../../src/images/airbnb-star.png",
          }
        }
        style={
          {
            "height": 40,
            "margin": 3,
            "tintColor": "#f1c40f",
            "transform": [
              {
                "scale": 1,
              },
            ],
            "width": 40,
          }
        }
        testID="RNVUI__Star-image"
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="RNVUI__Star"
    >
      <Image
        collapsable={false}
        source={
          {
            "testUri": "../../../src/images/airbnb-star.png",
          }
        }
        style={
          {
            "height": 40,
            "margin": 3,
            "tintColor": "#BDC3C7",
            "transform": [
              {
                "scale": 1,
              },
            ],
            "width": 40,
          }
        }
        testID="RNVUI__Star-image"
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="RNVUI__Star"
    >
      <Image
        collapsable={false}
        source={
          {
            "testUri": "../../../src/images/airbnb-star.png",
          }
        }
        style={
          {
            "height": 40,
            "margin": 3,
            "tintColor": "#BDC3C7",
            "transform": [
              {
                "scale": 1,
              },
            ],
            "width": 40,
          }
        }
        testID="RNVUI__Star-image"
      />
    </View>
  </View>
</View>
`;
